### ISSUE TITLE
Glyph table sort order

### ISSUE THREAD

**Issue by ahmedghazi** [graphicore/specimenTools#26]
Hi,
a client of mine asks me if i can manage the glyph table sort order, or at least have a glyph table per font variation (uppercase, lowercase, aåáâã, xyzåáâã).

**Comment by graphicore** [comment]
We could enable to inject some custom function via the options that allows selection and order of the glyphs that go into the table.
That way, you could completely customize the table and produce either of your suggested variants.

The signature would look something like this:

```js
/**
 * `this` is the GlyphTable.
 * returns: `glyphOrder` an array of glyph indexes 
 */
function customGlyphOrder( ) {
    var glyphOrder = [];
    this._font; // the opentype.js font object
    this._options; // the options can be used to modify the behavior
    ...
   return glyphOrder;
}
```

It could be also useful to allow some `data-` attributes scheme, which `GlyphTables` would have to know and pass down the values to  `GlyphTable`. Would be used like:

```html
<div class="glyph-table" data-glyphorder="lowercase"></div>
```

To interperete the `glyphorder` value `"lowercase"`would be the task of `customGlyphOrder`:

```js
function customGlyphOrder( ) {
    // this._options.glyphorder === 'lowercase';
}
```


What do you think?

**Comment by graphicore** [comment]
To give some context, the current default order would be created like this (compare https://github.com/graphicore/specimenTools/blob/master/lib/widgets/GlyphTable.js#L174):

```js
function customGlyphOrder( ) {
    var glyphOrder = [], i, l;
    for(i=0,l=this._font.glyphNames.names.length;i<l;i++)
        glyphOrder.push(i);
    return glyphOrder;
}
```

**Comment by ahmedghazi** [comment]
That is great, with the data attribute, the same way you do with de xHeight, select, order.

**Comment by graphicore** [comment]
> the same way you do with de xHeight, select, order.

What's the deal with `xHeight`?

For the sake of simplicity, I'd just provide one attribute `data-glyphorder` and you would have to parse it yourself, maybe pass JSON (caution with attribute quotes!):

```html
<div class="glyph-table" data-glyphorder='{select: "lowercase", order: "a-z"}'></div>
```

```js
function customGlyphOrder( ) {
    options = JSON.parse(this._options.glyphorder);
    // options.select === 'lowercase'
    // options.order === '"a-z'
}
```

Or a separated string:

```html
<div class="glyph-table" data-glyphorder="lowercase, a-z"></div>
```

```js
function customGlyphOrder( ) {
    options = this._options.glyphorder.split(',')
                  .map(function(str){return str.trim();})
    select = options[0]; // === 'lowercase'
    order = options[1]; // === 'a-z'
}
```

**Comment by ahmedghazi** [comment]
The best would be to filter through data attribute
uppercase
lowercase
small caps
the rest of the glyphs

**Comment by graphicore** [comment]
> The best would be to filter through data attribute

I'm suggesting that you'll implement this yourself. It's hard to get this right for all possible fonts. But, for one foundry, its probably feasible.

**Comment by ahmedghazi** [comment]
So in the "_initCells" function i can call the customGlyphOrder to change the order, but on the this._font.glyphNames.names i only have unicode, so i should implement depending on the range unicode?
Or is there an easier way?

**Comment by graphicore** [comment]
> So in the "_initCells" function i can call the customGlyphOrder to change the order,

Yes. It's not implemented yet, but you will be able to define your own `customGlyphOrder` function.

>  but on the this._font.glyphNames.names i only have unicode, so i should implement depending on the range unicode?

You can use all of `this._font` (opentypejs). Glyph names are likely "production names" like `uni1234`, so that's not much better than using unicodes, however, small caps may have a name like `uni1234.sc` but no own unicode. So that may help you to select all small caps and extract what their corresponding unicode is.
But this really depends on how the font is made, that's why I don't want to try to implement it generically.

**Comment by ahmedghazi** [comment]
Thx, i'll do that.

**Comment by graphicore** [comment]
Ok, I'll implement this.

What is the deal with "xHeight"?

**Comment by ahmedghazi** [comment]
The way you give parameters here https://graphicore.github.io/mdlFontSpecimen/
<div 
    class="mdlfs-x-height-diagram mdlfs-diagram" 
    data-diagram-name="xHeight" 
    data-diagram-ascent="800" 
    data-diagram-descent="5"></div>

**Comment by graphicore** [comment]
It's underway https://github.com/graphicore/specimenTools/pull/27

I modified the example/simple to test this, here's a gist:

https://gist.github.com/graphicore/5d8fe09b990ce20d10a4eac3301948dc

Notes:

index.html

Attributes starting with `data-glyphtable-` will be available in `this._dataAttributes.{name}`. E.g.:
 `data-glyphtable-contains` => `this._dataAttributes.contains`

```html
<div class="glyph-table" data-glyphtable-contains="A"></div>
```

main.js:

I'm only showing the interesting parts here.

`customGlyphOrder` is a bit silly but a good start.

```js
        function customGlyphOrder() {
            //jshint validthis:true
            var glyphOrder = [], i, l, name;
            for(i=0,l=this._font.glyphNames.names.length;i<l;i++) {
                name = this._font.glyphNames.names[i];
                if(name.indexOf(this._dataAttributes.contains) !== -1)
                    glyphOrder.push(i);
            }
            return glyphOrder;
        }

        var glyphTablesOptions = {
          glyphTable: {
            glyphOrderFunc: customGlyphOrder
          }
        };

        factories = [
        ...
          , ['glyph-table', GlyphTables, glyphTablesOptions]
        ...
        ];
```

**Comment by graphicore** [comment]
When you confirm that this is sufficient, I will merge. :-)
